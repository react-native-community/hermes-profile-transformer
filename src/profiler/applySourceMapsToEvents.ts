import path from 'path';
import { SourceMapConsumer, RawSourceMap } from 'source-map';
import { DurationEvent } from '../types/EventInterfaces';
import { SourceMap } from '../types/SourceMap';

/**
 * This function is a helper to the applySourceMapsToEvents. The node_module identification logic is implemented here based on the sourcemap url (if available). Incase a node_module could not be found, this defaults to the category of the event
 * @param defaultCategory The category the event is of by default without the use of Source maps
 * @param url The URL which can be parsed to interpret the new category of the event (depends on node_modules)
 */
const findNodeModuleNameIfExists = (
  defaultCategory: string,
  url: string | null
): string => {
  const obtainCategory = (url: string): string => {
    const dirs = url
      .substring(url.lastIndexOf(`${path.sep}node_modules${path.sep}`))
      .split(path.sep);
    return dirs.length > 2 && dirs[1] === 'node_modules'
      ? dirs[2]
      : defaultCategory;
  };
  return url ? obtainCategory(url) : defaultCategory;
};

/**
 * The unification of categories is important as we want identify the specific reasons why the application slows down, namely via unoptimised native/JS code, or react-native renders or third party modules. The common colours for node_modules can help idenitfy problems instantly
 * @param nodeModuleName The node module name associated with the event obtained via sourcemap, this nodeModule name is simply the output of @see findNodeModuleNameIfExists
 */
const improveCategories = (
  nodeModuleName: string,
  defaultCategory: string
): string => {
  // The nodeModuleName obtained from `findNodeModuleNameIfExists` by default is the original category name in the generated Hermes Profile. If we cannot isolate a nodeModule name, we simply return with the default category
  if (nodeModuleName === defaultCategory) {
    return defaultCategory;
  }
  // The events from these modules will fall under the umbrella of react-native events and hence be represented by the same colour
  const reactNativeModuleNames = ['react-native', 'react', 'metro'];
  if (reactNativeModuleNames.includes(nodeModuleName)) {
    return 'react-native-internals';
  } else {
    return 'other_node_modules';
  }
};

/**
 * Enhances the function line, column and params information and event categories
 * based on JavaScript source maps to make it easier to associate trace events with
 * the application code
 *
 * Throws error if args not set up in ChromeEvents
 * @param {SourceMap} sourceMap
 * @param {DurationEvent[]} chromeEvents
 * @param {string} indexBundleFileName
 * @throws If `args` for events are not populated
 * @returns {DurationEvent[]}
 */
const applySourceMapsToEvents = async (
  sourceMap: SourceMap,
  chromeEvents: DurationEvent[],
  indexBundleFileName: string | undefined
): Promise<DurationEvent[]> => {
  // SEE: Should file here be an optional parameter, so take indexBundleFileName as a parameter and use
  // a default name of `index.bundle`
  const rawSourceMap: RawSourceMap = {
    version: Number(sourceMap.version),
    file: indexBundleFileName || 'index.bundle',
    sources: sourceMap.sources,
    mappings: sourceMap.mappings,
    names: sourceMap.names,
  };

  const consumer = await new SourceMapConsumer(rawSourceMap);
  const events = chromeEvents.map((event: DurationEvent) => {
    if (event.args) {
      let line = Number(event.args.line);
      let column = Number(event.args.column);

      // For production profiles we construct line + column from the funcVirtAddr and offset
      // See https://github.com/facebook/metro/blob/139f58c1eeb7b61318f6307d5be650e5484ea1c5/packages/metro-symbolicate/src/Symbolication.js#L301
      if (
        event.args.funcVirtAddr !== undefined &&
        event.args.offset !== undefined
      ) {
        line = 1;
        column =
          Number(event.args.funcVirtAddr) + Number(event.args.offset) + 1;
      }

      const sm = consumer.originalPositionFor({
        line,
        column,
      });
      /**
       * The categories can help us better visualise the profile if we modify the categories.
       * We change these categories only in the root level and not deeper inside the args, just so we have our
       * original categories as well as these modified categories (as the modified categories simply help with visualisation)
       */
      const nodeModuleNameIfAvailable = findNodeModuleNameIfExists(
        event.cat!,
        sm.source
      );

      event.cat = improveCategories(nodeModuleNameIfAvailable, event.cat!);
      event.args = {
        ...event.args,
        url: sm.source,
        line: sm.line,
        column: sm.column,
        params: sm.name,
        allocatedCategory: event.cat,
        allocatedName: event.name,
        node_module: nodeModuleNameIfAvailable,
      };
    } else {
      throw new Error(
        `Source maps could not be derived for an event at ${event.ts} and with stackFrame ID ${event.sf}`
      );
    }
    return event;
  });
  consumer.destroy();
  return events;
};

export default applySourceMapsToEvents;
