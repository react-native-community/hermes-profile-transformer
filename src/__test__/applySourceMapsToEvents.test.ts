import { SourceMap } from '../types/SourceMap';
import { SourceMapConsumer } from 'source-map';
import { DurationEvent } from '../types/EventInterfaces';
import { EventsPhase } from '../types/Phases';
import applySourceMapsToEvents from '../profiler/applySourceMapsToEvents';

const mockOriginalPositionFor = jest.fn();

jest.mock('source-map', () => ({
  SourceMapConsumer: jest.fn().mockImplementation(() => ({
    originalPositionFor: mockOriginalPositionFor.mockReturnValue({
      source: 'source.js',
      line: 1,
      column: 10,
      name: 'console.log',
    }),
    destroy: jest.fn(),
  })),
}));

describe('applySourceMapsToEvents', () => {
  let defaultEvents: DurationEvent[] = [];

  const mockSourceMap: SourceMap = {
    version: '3',
    sources: ['source.js'],
    sourceContent: ['console.log("Hello, World!");'],
    x_facebook_sources: null,
    names: ['console', 'log', 'Hello, World!'],
    mappings: 'AAAAA,QAAAC,IAAA',
  };

  beforeEach(() => {
    ((SourceMapConsumer as unknown) as jest.Mock).mockImplementation(() => ({
      originalPositionFor: mockOriginalPositionFor.mockReturnValue({
        source: 'source.js',
        line: 5,
        column: 10,
        name: 'console.log',
      }),
      destroy: jest.fn(),
    }));
    mockOriginalPositionFor.mockClear();

    // applySourceMapsToEvents modifies the events array, so we need to reset it before each test
    defaultEvents = [
      {
        ph: EventsPhase.DURATION_EVENTS_BEGIN,
        args: {
          line: 1,
          column: 1,
        },
        cat: 'default-category',
        name: 'event1',
        ts: 1000,
        pid: 1,
        tid: 1,
      },
      {
        ph: EventsPhase.DURATION_EVENTS_END,
        args: {
          line: 1,
          column: 1,
        },
        cat: 'default-category',
        name: 'event1',
        ts: 1010,
        pid: 1,
        tid: 1,
      },
    ];
  });

  describe('dev profiles', () => {
    it('should enhance events with source map information', async () => {
      const indexBundleFileName = 'index.bundle';
      const enhancedEvents = await applySourceMapsToEvents(
        mockSourceMap,
        defaultEvents,
        indexBundleFileName
      );

      expect(enhancedEvents).toEqual([
        expect.objectContaining({
          args: {
            url: 'source.js',
            line: 5,
            column: 10,
            params: 'console.log',
            allocatedCategory: 'default-category',
            allocatedName: 'event1',
            node_module: 'default-category',
          },
        }),
        expect.objectContaining({
          args: {
            url: 'source.js',
            line: 5,
            column: 10,
            params: 'console.log',
            allocatedCategory: 'default-category',
            allocatedName: 'event1',
            node_module: 'default-category',
          },
        }),
      ]);
    });

    it('should throw an error if args are not populated', async () => {
      const eventsWithMissingArgs: DurationEvent[] = [
        {
          ...defaultEvents[0],
          args: undefined as any,
        },
      ];

      await expect(
        applySourceMapsToEvents(mockSourceMap, eventsWithMissingArgs, undefined)
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('production profiles', () => {
    it('should enhance events with source map information', async () => {
      const productionEvents: DurationEvent[] = [
        {
          ph: EventsPhase.DURATION_EVENTS_BEGIN,
          args: {
            funcVirtAddr: '5',
            offset: '7',
          },
          cat: 'default-category',
          name: 'event1',
          ts: 1000,
          pid: 1,
          tid: 1,
        },
        {
          ph: EventsPhase.DURATION_EVENTS_END,
          args: {
            funcVirtAddr: '5',
            offset: '7',
          },
          cat: 'default-category',
          name: 'event1',
          ts: 1010,
          pid: 1,
          tid: 1,
        },
      ];

      const indexBundleFileName = 'index.bundle';
      const enhancedEvents = await applySourceMapsToEvents(
        mockSourceMap,
        productionEvents,
        indexBundleFileName
      );

      expect(mockOriginalPositionFor).toHaveBeenCalledTimes(2);
      expect(mockOriginalPositionFor).toHaveBeenCalledWith({
        line: 1,
        column: 13,
      });

      expect(enhancedEvents).toEqual([
        expect.objectContaining({
          args: {
            url: 'source.js',
            line: 5,
            column: 10,
            funcVirtAddr: '5',
            offset: '7',
            params: 'console.log',
            allocatedCategory: 'default-category',
            allocatedName: 'event1',
            node_module: 'default-category',
          },
        }),
        expect.objectContaining({
          args: {
            url: 'source.js',
            line: 5,
            column: 10,
            funcVirtAddr: '5',
            offset: '7',
            params: 'console.log',
            allocatedCategory: 'default-category',
            allocatedName: 'event1',
            node_module: 'default-category',
          },
        }),
      ]);
    });
  });

  describe('setting event category', () => {
    it.each([
      {
        nodeModuleName: 'react-native',
        expectedCategory: 'react-native-internals',
      },
      { nodeModuleName: 'react', expectedCategory: 'react-native-internals' },
      { nodeModuleName: 'metro', expectedCategory: 'react-native-internals' },
      {
        nodeModuleName: 'other-module',
        expectedCategory: 'other_node_modules',
      },
    ])(
      'should correctly improve categories for node modules',
      async ({ nodeModuleName, expectedCategory }) => {
        ((SourceMapConsumer as unknown) as jest.Mock).mockImplementation(
          () => ({
            originalPositionFor: jest.fn().mockReturnValue({
              source: `/workdir/node_modules/${nodeModuleName}/index.js`,
              line: 1,
              column: 10,
              name: 'console.log',
            }),
            destroy: jest.fn(),
          })
        );

        const enhancedEvents = await applySourceMapsToEvents(
          mockSourceMap,
          defaultEvents,
          undefined
        );

        expect(enhancedEvents).toEqual(
          expect.arrayContaining([
            expect.objectContaining({
              cat: expectedCategory,
            }),
          ])
        );
      }
    );

    it('should default to the source as the category', async () => {
      ((SourceMapConsumer as unknown) as jest.Mock).mockImplementation(() => ({
        originalPositionFor: jest.fn().mockReturnValue({
          source: `asdfasd.js`,
          line: 1,
          column: 10,
          name: 'console.log',
        }),
        destroy: jest.fn(),
      }));

      const enhancedEvents = await applySourceMapsToEvents(
        mockSourceMap,
        defaultEvents,
        undefined
      );

      expect(enhancedEvents).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            cat: 'default-category',
          }),
        ])
      );
    });
  });
});
