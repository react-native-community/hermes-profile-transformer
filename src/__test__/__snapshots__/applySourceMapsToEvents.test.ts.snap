// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applySourceMapsToEvents dev profiles should throw an error if args are not populated 1`] = `"Source maps could not be derived for an event at 1000 and with stackFrame ID undefined"`;
